### PR TITLE
pulling user compliance pdf file

### PR DESCRIPTION
**Pull request by eoliveros**
- [x] retrieve the user compliance file from back2blaze

### CODE REVIEW

File: models.py [hunk 1]
```diff
@@ -778,6 +778,10 @@ def url(self):
     def count(cls, session):
         return session.query(cls).count()
 
+    @classmethod
+    def from_user_id(cls, session, user_id):
```
Comment (djpnewton) [inline models.py:782]: We have an established pattern of `from_user`
```suggestion
    def from_user(cls, session, user):
        return session.query(cls).filter(cls.user_id == user.id).first()
```

File: templates/index.html [hunk 2]
```diff
@@ -24,20 +24,25 @@
...
-                    <p><center>Test Websocket</center></p>
-                </div>
-                </a>
+            <div class='dashboard'>
+                <p><center>Test Websocket</center></p>
+            </div>
+            </a>
+            <a href="{{ url_for('user_compliance') }}">
+            <div class='dashboard'>
+                <p><center>User Compliance</center></p>
```
Comment (djpnewton) [inline templates/index.html:43]: Lets keep the verbiage consistent within the app

```suggestion
                <p><center>User KYC</center></p>
```

File: web.py [hunk 3]
```diff
@@ -213,6 +214,27 @@ def test_ws():
             flash('Event not yet implemented', 'danger')
     return render_template('test_ws.html', recipient=recipient, event=event, events=events)
 
+@roles_accepted(Role.ROLE_ADMIN, Role.ROLE_FINANCE)
+@app.route('/user_compliance', methods=['GET', 'POST'])
```
Comment (djpnewton) [inline web.py:218]:
```suggestion
@app.route('/user_kyc', methods=['GET', 'POST'])
```

File: web.py [hunk 4]
```diff
@@ -213,6 +214,27 @@ def test_ws():
...
+def user_compliance():
```
Comment (djpnewton) [inline web.py:219]:
```suggestion
def user_kyc():
```

File: web.py [hunk 5]
```diff
@@ -213,6 +214,27 @@ def test_ws():
...
+    email = ''
+    if request.method == 'POST':
+        email = request.form['email']
+        user = User.from_email(db.session, email)
+        if user:
+            kycrequest = KycRequest.from_user_id(db.session, user.id)
+            if kycrequest:
+                token = kycrequest.token
+                pdf = kyc_core.download_pdf_backup(token)
+                if not pdf:
```
Comment (djpnewton) [inline web.py:229]: nicer error using `flash()`
```suggestion
                if pdf:
                    response = make_response(pdf)
                    response.headers.set('Content-Type', 'application/pdf')
                    return response
                else:
                   flash('failed to download pdf', 'danger')
```

File: web.py [hunk 6]
```diff
@@ -213,6 +214,27 @@ def test_ws():
...
+                    return 'failed to download pdf'
+                response = make_response(pdf)
+                response.headers.set('Content-Type', 'application/pdf')
+                return response
+        else:
+            flash('User/compliance not found', 'danger')
```
Comment (djpnewton) [inline web.py:235]:
```suggestion
            flash('User not found', 'danger')
```

**Review by djpnewton — CHANGES_REQUESTED** [review]

File: models.py (hunk 1 above)
Comment (eoliveros) [reply models.py:782]: changed to this:

def from_user(cls, session, user):
        return session.query(cls).filter(and_(cls.user_id == user.id, cls.status == 'completed')).first()

File: templates/index.html (hunk 2 above)
Comment (eoliveros) [reply templates/index.html:43]: changed compliance to kyc

File: web.py (hunk 3 above)
Comment (eoliveros) [reply web.py:218]: changed this link

File: web.py (hunk 4 above)
Comment (eoliveros) [reply web.py:219]: changed the _compliance to _kyc

File: web.py (hunk 5 above)
Comment (eoliveros) [reply web.py:229]: committed the change above

File: web.py (hunk 6 above)
Comment (eoliveros) [reply web.py:235]: changed this as well.

File: web.py [hunk 7]
```diff
@@ -213,6 +214,32 @@ def test_ws():
...
+            if kycrequest.status == 'completed':
+                token = kycrequest.token
+                pdf = kyc_core.download_pdf_backup(token)
+                if pdf:
+                    response = make_response(pdf)
+                    response.headers.set('Content-Type', 'application/pdf')
+                    return response
+                else:
+                    flash('failed to download pdf', 'danger')
+            elif kycrequest.status == 'created':
```
Comment (djpnewton) [inline web.py:235]: dont use a string literal use the constant `STATUS_CREATED`

File: web.py [hunk 8]
```diff
@@ -213,6 +214,32 @@ def test_ws():
...
+@roles_accepted(Role.ROLE_ADMIN, Role.ROLE_FINANCE)
+@app.route('/user_kyc', methods=['GET', 'POST'])
+def user_kyc():
+    email = ''
+    if request.method == 'POST':
+        email = request.form['email']
+        user = User.from_email(db.session, email)
+        if user:
+            kycrequest = KycRequest.from_user(db.session, user)
+            if kycrequest.status == 'completed':
```
Comment (djpnewton) [inline web.py:226]: dont use a string literal use the constant `STATUS_COMPETED`

**Review by djpnewton — CHANGES_REQUESTED** [review]

File: web.py (hunk 7 above)
Comment (eoliveros) [reply web.py]: changed to the above

File: web.py (hunk 8 above)
Comment (eoliveros) [reply web.py]: changed as per above

File: web.py [hunk 9]
```diff
@@ -213,6 +214,33 @@ def test_ws():
...
+        if user:
+            kycrequest = KycRequest.from_user(db.session, user)
+            if kycrequest.status == kycrequest.STATUS_COMPLETED:
+                token = kycrequest.token
+                pdf = kyc_core.download_pdf_backup(token)
+                if pdf:
+                    response = make_response(pdf)
+                    response.headers.set('Content-Type', 'application/pdf')
+                    return response
+                #else:
```
Comment (djpnewton) [inline web.py:233]: remove dead code

**Review by djpnewton — CHANGES_REQUESTED** [review]
is it working?

File: web.py (hunk 9 above)
Comment (eoliveros) [reply web.py]: apart from the file being downloaded... everything bout it seems to be working.

Comment (eoliveros) [reply web.py]: changed some of the host var more specifically the 
```
- b2_account_id 
- b2_application_key
- kyc_bucket
```

but now im getting the following:
```
  File "/home/eoliveros/git/beryllium/venv/lib/python3.9/site-packages/flask/helpers.py", line 191, in make_response
    return current_app.make_response(args)
  File "/home/eoliveros/git/beryllium/venv/lib/python3.9/site-packages/flask/app.py", line 1727, in make_response
    raise TypeError(
TypeError: The view function did not return a valid response. The return type must be a string, dict, tuple, Response instance, or WSGI callable, but it was a BytesIO.
[geventwebsocket.handler INFO] 172.23.96.1 - - [2021-09-30 16:33:41] "POST /user_kyc HTTP/1.1" 500 542 1.840609
```

not sure what this message means.
```

**Comment by eoliveros** [comment]
replacing part of the code:
```
                if pdf:
                    response = make_response(pdf)
                    response.headers.set('Content-Type', 'application/pdf')
                    return response
```

with this:
```
                if pdf:
                    filename = f"{email}.pdf"
                    return Response(
                        pdf,
                        mimetype="application/pdf",
                        headers={"Content-disposition":
                                 f"attachment; filename={filename}"})
```

This way when it successfully finds the file the user gets a box to "Save" the file locally.

**Comment by eoliveros** [comment]
can successfully download my test pdf file from backblaze2.

File: web.py [hunk 10]
```diff
@@ -213,6 +214,34 @@ def test_ws():
...
+            if kycrequest.status == kycrequest.STATUS_COMPLETED:
+                token = kycrequest.token
+                pdf = kyc_core.download_pdf_backup(token)
+                if pdf:
+                    filename = f"{email}.pdf"
+                    return Response(
+                        pdf,
+                        mimetype="application/pdf",
+                        headers={"Content-disposition":
+                                 f"attachment; filename={filename}"})
```
Comment (djpnewton) [inline web.py:235]:
```suggestion
                                 f"attachment; filename={email}.pdf"})
```

File: web.py [hunk 11]
```diff
@@ -213,6 +214,34 @@ def test_ws():
...
+    if request.method == 'POST':
+        email = request.form['email']
+        user = User.from_email(db.session, email)
+        if user:
+            kycrequest = KycRequest.from_user(db.session, user)
+            if kycrequest.status == kycrequest.STATUS_COMPLETED:
+                token = kycrequest.token
+                pdf = kyc_core.download_pdf_backup(token)
+                if pdf:
+                    filename = f"{email}.pdf"
```
Comment (djpnewton) [inline web.py:230]:
```suggestion

```

**Review by djpnewton — CHANGES_REQUESTED** [review]
The rest of the file uses single quotes `'` so replace the double quotes you added in web.py with single quotes for string literals

File: web.py (hunk 10 above)
Comment (eoliveros) [reply web.py:235]: changed {filename} to {email}.pdf

File: web.py (hunk 11 above)
Comment (eoliveros) [reply web.py:230]: removed the line.

File: web.py [hunk 12]
```diff
@@ -213,6 +214,33 @@ def test_ws():
             flash('Event not yet implemented', 'danger')
     return render_template('test_ws.html', recipient=recipient, event=event, events=events)
 
+@roles_accepted(Role.ROLE_ADMIN, Role.ROLE_FINANCE)
+@app.route('/user_kyc', methods=['GET', 'POST'])
+def user_kyc():
+    email = ''
+    if request.method == 'POST':
+        email = request.form['email']
```
Comment (djpnewton) [inline web.py:222]:
```suggestion
        email = request.form['email']
        if not email:
            flash('please enter an email address', 'danger')
            return render_template('user_kyc.html')
        email = email.lower()
```

**Pull request merged**